modules/distributor/usage/tracker.go

```diff
@@ -0,0 +1,423 @@
+		sanitizedDimensions := make(map[string]string, len(dimensions))
+		for k, v := range dimensions {
+			// Get the final sanitized output label for this
+			// dimension.  Dimensions are key-value pairs with
+			// optional value.  If value is empty string, then
+			// we use the just the key.  Regardless the output
+			// is always the sanitized version.
+			// Example:
+			//    service.name="" 			=> "service_name"
+			//    service.name="foo.bar"	=> "foo_bar"
+			var sanitized string
+			if v == "" {
+				// The dimension is using default mapping
+				v = k
+			}
+			sanitized = strutil.SanitizeFullLabelName(v)
+			sanitizedDimensions[k] = sanitized
+		}
+
+		// Step 2
+		// Build the final list of sorted/distinct outputs
+		t.sortedKeys = t.sortedKeys[:0]
+		for _, v := range sanitizedDimensions {
+			if !slices.Contains(t.sortedKeys, v) {
+				t.sortedKeys = append(t.sortedKeys, v)
+			}
+		}
+		slices.Sort(t.sortedKeys)
+
+		// Step 3
+		// Prepare the mapping from raw attribute names to the final location of
+		// where it goes in the output buffers. This avoids another layer of indirection.
+		t.mapping = t.mapping[:0]
+		for k := range dimensions {
+			i := slices.Index(t.sortedKeys, sanitizedDimensions[k])
+			t.mapping = append(t.mapping, mapping{
+				from: k,
+				to:   i,
+			})
+		}
+
+		// Step 4
+		// Prepopulate the buffers and precompute the overflow bucket
+		t.buffer1 = make([]string, len(t.sortedKeys))
+		t.buffer2 = make([]string, len(t.sortedKeys))
+		t.buffer3 = make([]string, len(t.sortedKeys))
+		for i := range t.sortedKeys {
+			t.buffer1[i] = overflowLabel
+		}
+		t.overflow = hash(t.sortedKeys, t.buffer1)
+	}
+	return t.mapping, t.buffer1, t.buffer2, t.buffer3
+}
+
+// func (t *tenantUsage) getSeries(labels, values []string, maxCardinality uint64) *bucket {
+func (t *tenantUsage) getSeries(buffer []string, maxCardinality uint64) *bucket {
+	h := hash(t.sortedKeys, buffer)
+
+	b := t.series[h]
+	if b == nil {
+		// Before creating a new series, check for cardinality limit.
+		if uint64(len(t.series)) >= maxCardinality {
+			// Overflow
+			// This tenant is at the maximum number of series.  In this case all data
+			// goes into the final overflow bucket. It has the same dimensions as the
+			// current configuration, except every label is overridden to the special overflow value.
+			for k := range buffer {
+				buffer[k] = overflowLabel
+			}
+			h = t.overflow
+			b = t.series[h]
+		}
+	}
+
+	if b == nil {
+		// First encounter with this series. Initialize it.
+		// Detach a copy of the values
+		v := make([]string, len(buffer))
+		copy(v, buffer)
+		b = &bucket{
+			// Metric description - constant for this pass now that the dimensions are known
+			descr:  prometheus.NewDesc("tempo_usage_tracker_bytes_received_total", "bytes total received with these attributes", t.sortedKeys, t.constLabels),
+			labels: v,
+		}
+		t.series[h] = b
+	}
+	return b
+}
+
+type Tracker struct {
+	mtx      sync.Mutex
```

Comment:
Is the mutex required for another concurrent operation besides getting the tenantUsage? Maybe we could use a sync.Map instead? It should be faster b/c this is a read-heavy

Reply:
Yes the mutex is needed to handle other concurrent pushes received by the distributor.  I actually expect it to be write-heavy.  All received traffic will be written to the metrics, and scraping will be less frequent (15s or 1m).   

Regarding sync.Map, we've tried it in a few areas and have found plain map + mutex to provide better performance. The ingester live-traces is good example.   I think it's the case where the usage doesn't fit the optimal cases described in the [docs](https://pkg.go.dev/sync#Map):  read-heavy, or writing disjoint sets of keys.  In this case I expect it to be write-heavy, and each tenant is a separate map, so is generally writing the same keys.

If this is shown to be a bottleneck, another option would be 2-layer mutexes, one to guard the tenant map, and then another per-tenant mutex within.  Generators are using this approach: [registry mutex](https://github.com/grafana/tempo/blob/main/modules/generator/registry/registry.go#L66) and [series mutex](https://github.com/grafana/tempo/blob/main/modules/generator/registry/histogram.go#L27)

Reply:
Thank you Marty, your explanation has been very enlightening